Update types of some brig endpoints to be federation-aware

mheinzel opened the pull request:
One thing I'm still thinking about (and left some TODOs for):

In Galley when encounter a remote ID where we can't handle it yet, we throw a HTTP 501 "federation-not-implemented". Here, it's not as easy since we throw errors into different sum types for different routes (ClientError, UserError, ConnectionError), which we would all have to extend with a case for federation. I'm currently leaning towards just doing that (instead of just treating remote IDs as non-existent ones).

arianvp commented on services/brig/src/Brig/API.hs:
```diff
@@ -1133,30 +1157,33 @@ getUserNameH (_ ::: self) = do
     Nothing -> setStatus status404 empty
 
-listUsersH :: JSON ::: UserId ::: Either (List UserId) (List Handle) -> Handler Response
+listUsersH :: JSON ::: UserId ::: Either (List OpaqueUserId) (List Handle) -> Handler Response
```
Should the handles be qualified here?

mheinzel replied:
Optionally qualified, right? Good catch.

I don't think we are using this anywhere right now (we are definitely not in web) and I'm not sure what the use case is, but I'll fix it.

mheinzel replied:
done

arianvp commented on libs/types-common/src/Data/IdMapping.hs:
```diff
@@ -17,6 +17,11 @@ opaqueIdFromMappedOrLocal = \case
 
+partitionMappedOrLocalIds :: Foldable f => f (MappedOrLocalId a) -> ([Id a], [IdMapping a])
+partitionMappedOrLocalIds = foldMap $ \case
+  Mapped mapping -> (mempty, [mapping])
```
neat

arianvp commented on libs/types-common/src/Data/Qualified.hs:
```diff
@@ -4,18 +4,54 @@ module Data.Qualified where
+Either is a hack, conceptually we want this:
+
+data OptionallyQualified a
+  = OptionallyQualified
```
Do you still want to resolve htis before merging?

mheinzel replied:
Yes, that's the main things that's missing.

mheinzel replied:
Implemented now.

arianvp commented on services/brig/src/Brig/API/Client.hs:
```diff
@@ -101,17 +124,33 @@ rmClient u con clt pw =
+  Local localUser ->
+    claimLocalPrekey localUser c
+  Mapped _ ->
+    -- TODO: error?
```
no not an error right? we contact the other backend and ask it to claim a prekey for us no?

mheinzel replied:
Only until we implement the logic. In Galley, we fail with some `federation-not-implemented` in these cases (a remote ID being used, but the function not being prepared for it).

mheinzel replied:
I removed all of these now. Still not 100% happy with it, but seems reasonable, since it will just behave the same as if a local user was not found.

arianvp commented on services/brig/src/Brig/API/Client.hs:
```diff
@@ -101,17 +124,33 @@ rmClient u con clt pw =
+  Local localUser ->
+    claimLocalPrekeyBundle localUser
+  Mapped IdMapping {idMappingLocal} ->
+    -- TODO: error?
```
No, same as above

mheinzel commented:
I rebases and implemented a parser for `OptionallyQualified Handle` and IDs.

To do that, I also rewrote the parsers for handles, IDs and domains, to make them composable (which they aren't if they always try to consume the whole input). Then I based all other validation and smart constructors on these Attoparsec parsers. I added some roundtrip tests, but it would still be nice if someone could have a look at these parsers specifically.

arianvp commented on libs/types-common/src/Data/Domain.hs:
```diff
@@ -43,12 +74,52 @@ instance FromJSON Domain where
+          `QC.suchThat` ((<= 255) . Text.length)
+      -- <label> ::= <let-dig> [ [ <ldh-str> ] <let-dig> ]
+      label =
+        conc [letDig, opt (conc [opt ldhStr, letDig])]
```
We had this cool parser combinator framework at UU that would  generate a parser, pretty printer, and quickcheck generator for each  AST that you defined.  I miss it now.  (As this is just the same as the parser code, but in a different applicative :P; wonder if we can abuse that fact somehow (Free applicative?) )

Ohwell not that important; just an observation

mheinzel replied:
I remember seeing something like that and yes, it's nice.

The parser doesn't adhere to the grammar as closely as the generator here, so testing them against each other gives me some confidence. Although I only know the parser is not too restrictive, but it could still be too permissive. :thinking:

fisx commented on libs/types-common/src/Data/Domain.hs:
```diff
@@ -20,20 +37,34 @@ domainText :: Domain -> Text
+  parser = domainParser
+
+domainParser :: Atto.Parser Domain
+domainParser = do
```
heh, you couldn't help yourself.  :)

(i assume this is because the domain validation in the email package wasn't as thorough as this one.)

mheinzel replied:
Yes, it allows some extra things like comments (`(comment)example.com`) and bracketed IP addresses (`[1.1.1.1.]`).

mheinzel replied:
We now also decided to stick to FQDNs, requiring at least two labels, so I'll update the parser again.

mheinzel replied:
done.

fisx commented on libs/types-common/src/Data/Id.hs:
```diff
@@ -112,20 +114,30 @@ newtype Id a
-    x <- takeByteString
-    case fromASCIIBytes x of
+    x <-
+      -- we only want the matching ByteString
```
lacking context here.  do you mean the matching *part* of the bytestring?  if so that sounds odd to me.

also, one-letter variable name?  (sorry :))

mheinzel replied:
Yes, only the matching part. This is because if we compose the parser with e.g. the domain parser to parse a qualified ID like `61a73a52-e526-4892-82a9-3d638d77629f@example.com`, the parser for the UUID should not consume that whole string, but leave the `@example.com` for the next parser.

I'll add that explanation.

fisx commented on libs/types-common/test/Test/Domain.hs:
```diff
@@ -0,0 +1,27 @@
+
+testDomainSerialization :: [TestTree]
+testDomainSerialization =
+  [ testProperty "Arbitrary DomainText generates valid domains" $
```
It's good to phrase it like this, and not "mkDomain works for *all* inputs".  :)

fisx replied:
You could run `mkDomain` on some invalid (and valid) domains given as string literals, if you really wanted to.

mheinzel replied:
I'll do that, similar to the tests for handles.

mheinzel replied:
added the tests, done.

arianvp commented on libs/types-common/src/Data/Handle.hs:
```diff
@@ -23,37 +31,43 @@ newtype Handle
+handleParser :: Atto.Parser Handle
+handleParser = do
+  bs <- matching (Atto.count 2 handleChar *> Atto.skipMany handleChar)
+  when (BS.length bs > 256) $
```
Why not use count here like before?

mheinzel replied:
No fundamental reason. `count n (optional _)` seems like it would always spin `n` times, even if there is no valid handleChar anymore. In addition, it will allocate a bunch of unnecessary `Maybe`s. Maybe the error message is also clearer? On the other hand, `skipMany` will not stop after 256 chars, so not sure which one is better. :man_shrugging:

arianvp commented on libs/types-common/src/Data/Id.hs:
```diff
@@ -112,20 +114,30 @@ newtype Id a
       Nothing -> fail "Invalid UUID"
       Just ui -> return (Id ui)
+    where
+      matching = fmap fst . Atto.match
```
TIL Atto.match;  I did some nasty things in hscim instead.

mheinzel commented:
All feedback has been incorporated and we require multiple domain labels now (not allowing dotless domains).

I'm rebasing now. Anyone wants to take another look in the meantime?

mheinzel commented:
I rebase and needed some changes since the cabal files have now been checked in. If CI passed, I'll merge.
